lib/nostrum/shard/session.ex

```diff
@@ -81,59 +110,76 @@ defmodule Nostrum.Shard.Session do
 
     case from_handle do
       {new_state, reply} ->
-        {:reply, {:binary, reply}, new_state}
+        :ok = :gun.ws_send(state.conn, {:binary, reply})
+        {:noreply, new_state}
 
       new_state ->
-        {:ok, new_state}
+        {:noreply, new_state}
     end
   end
 
-  def handle_cast({:status_update, payload}, state) do
-    {:reply, {:binary, payload}, state}
-  end
+  def handle_info(
+        {:gun_down, conn, protocol, reason, _killed_streams, _unprocessed_streams},
+        state
+      ) do
+    Logger.warn(fn -> "connection down (#{protocol} #{reason}), attempting reconnect" end)
 
-  def handle_cast({:request_guild_members, payload}, state) do
-    {:reply, {:binary, payload}, state}
-  end
-
-  def handle_cast(:heartbeat, %{heartbeat_ack: false} = state) do
-    Logger.warn("heartbeat_ack not received in time, disconnecting")
-    {:close, state}
-  end
+    {:ok, :cancel} = :timer.cancel(state.heartbeat_ref)
+    :ok = :gun.close(conn)
 
-  def handle_cast(:heartbeat, state) do
-    {:ok, ref} =
-      :timer.apply_after(state.heartbeat_interval, WebSockex, :cast, [state.conn_pid, :heartbeat])
+    worker = connect(state.gateway)
 
-    {:reply, {:binary, Payload.heartbeat_payload(state.seq)},
-     %{state | heartbeat_ref: ref, heartbeat_ack: false, last_heartbeat_send: DateTime.utc_now()}}
+    {:noreply, %{state | conn: worker}}
   end
 
-  def handle_disconnect(%{reason: reason}, state) when is_tuple(reason) do
+  def handle_info({:gun_ws, conn, _stream, {:close, errno, reason}}, state) do
     Logger.warn(fn ->
-      "websocket disconnected with reason #{inspect(reason)}, attempting reconnect"
+      "websocket disconnected with code #{errno}, reason #{inspect(reason)}, " <>
+        "attempting reconnect"
```

Comment:
Maybe we should move this "attempting reconnect" portion elsewhere if we're not going to actually try reconnecting in this callback.